bin/propolis-server/src/lib/migrate/compat.rs

```diff
@@ -0,0 +1,791 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Checks for compatibility of two instance specs.
+
+use std::collections::HashMap;
+
+use crate::spec::{self, SerialPortDevice};
+
+use propolis_api_types::instance_spec::{
+    components::{
+        board::Chipset,
+        devices::{PciPciBridge, SerialPortNumber},
+    },
+    PciPath,
+};
+use thiserror::Error;
+
+trait CompatComponent {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), ComponentIncompatibility>;
+}
+
+trait CompatCollection {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), CollectionIncompatibility>;
+}
+
+#[derive(Debug, Error)]
+pub enum CompatibilityError {
+    #[error("specs have incompatible boards")]
+    Board(#[from] BoardIncompatibility),
+
+    #[error("specs have incompatible pvpanic settings")]
+    Pvpanic(#[from] PvpanicIncompatibility),
+
+    #[error("collection {0} incompatible")]
+    Collection(String, #[source] CollectionIncompatibility),
+
+    #[cfg(feature = "falcon")]
+    #[error("can't migrate instances containing softnpu devices")]
+    SoftNpu,
+}
+
+#[derive(Debug, Error)]
+pub enum BoardIncompatibility {
+    #[error("boards have different CPU counts (self: {this}, other: {other})")]
+    CpuCount { this: u8, other: u8 },
+
+    #[error(
+        "boards have different memory sizes (self: {this}, other: {other})"
+    )]
+    MemorySize { this: u64, other: u64 },
+
+    #[error(
+        "chipsets have different PCIe settings (self: {this}, other: {other})"
+    )]
+    PcieEnabled { this: bool, other: bool },
+}
+
+#[derive(Debug, Error)]
+pub enum DiskIncompatibility {
+    #[error(
+        "disks have different device interfaces (self: {this}, other: {other})"
+    )]
+    Interface { this: &'static str, other: &'static str },
+
+    #[error("disks have different PCI paths (self: {this}, other: {other})")]
+    PciPath { this: PciPath, other: PciPath },
+
+    #[error(
+        "disks have different backend names (self: {this}, other: {other})"
+    )]
+    BackendName { this: String, other: String },
+
+    #[error(
+        "disks have different backend kinds (self: {this}, other: {other})"
+    )]
+    BackendKind { this: &'static str, other: &'static str },
+
+    #[error(
+        "disks have different read-only settings (self: {this}, other: {other})"
+    )]
+    ReadOnly { this: bool, other: bool },
+}
+
+#[derive(Debug, Error)]
+pub enum NicIncompatibility {
+    #[error("NICs have different PCI paths (self: {this}, other: {other})")]
+    PciPath { this: PciPath, other: PciPath },
+
+    #[error(
+        "NICs have different backend names (self: {this}, other: {other})"
```

Comment:
Same thing as above, these should maybe be `Debug`:
```suggestion
        "NICs have different backend names (self: {this:?}, other: {other:?})"
```